src/Helper/MultiLockHandler.php

```diff
@@ -0,0 +1,78 @@
+<?php
+
+namespace Consul\Helper;
+
+use Consul\Services\KV;
+use Consul\Services\Session;
+
+class MultiLockHandler
+{
+    private array $resources;
+    private int $ttl;
+    private Session $session;
+    private KV $kv;
+    private string $sessionId;
+    private string $lockPath;
+
+    public function __construct(array $resources, int $ttl, Session $session, KV $kv, $lockPath)
+    {
+        $this->resources = $resources;
+        $this->ttl = $ttl;
+        $this->session = $session;
+        $this->kv = $kv;
+        $this->lockPath = $lockPath;
+    }
+
+    public function lock(): bool
+    {
+        $result = true;
+
+        // Start a session
+        $this->sessionId = $this->session->create(['LockDelay' => 0, 'TTL' => "{$this->ttl}s"])->json()['ID'];
+
+        $lockedResources = [];
+
+        foreach ($this->resources as $resource) {
+            // Lock a key / value with the current session
+            $lockAcquired = $this->kv->put($this->lockPath.$resource, '', ['acquire' => $this->sessionId])->json();
+
+            if (false === $lockAcquired) {
+                $result = false;
+
+                break;
+            }
+
+            $lockedResources[] = $resource;
+        }
+
+        if (!$result) {
+            $this->releaseResources($lockedResources);
+        }
```

Comment:
I would use a `try { } finally { if (!$result) {$this->releaseResources($lockedResources); }`

to ensure you remove everything when something goes wrong.

Reply:
done